### ISSUE TITLE
Media, MediaList, MediaListPlayer unsatisfactory implementation/API

### ISSUE THREAD

**Issue by caprica** [caprica/vlcj#754]
There are some outstanding issues with regards to Media, MediaLists etc.

There is an awkwardness that sometimes the media player "owns" a Media instance, and sometimes it is the client application.

MediaList needs special attention, in general and specifically for events. The subItemTreeAdded and subItemAdded currently pass libvlc_media_t on their listener method, should this be wrapped in a Media object?

Need to review the whole Media/MediaList shebang with particular attention to who "owns" the media, what is retained/released natively and what must be released by vlcj or the client application.

Also, when pulling a Media instance out of a MediaList do we want to return the same Media instance each time (the same that was added) or a new instance? Use-cases are unclear.

The CueTest test/demo class is a good test for the MediaList, currently it hard-crashes when trying to play a particular sub-item.

When using the DefaultMediaPlayer, the MediaList instance can be null even if there are sub-items created.

**Comment by caprica** [comment]
The reason for the hard crash...

If the option was not set on the media player to auto play sub-items, then the sub-items list was never set on the associated media list player. Consequently crash.

It looks like auto playing sub-items should no longer be an option, if we use media player and an associated media list player correctly, the sub-item will just start (likely always what you want anyway) and so we should always just set the list.

Note that setting the list on the associated media list player starts playing.

It would actually be really nice to get rid of the API to auto play sub-items and just have it be the natural behaviour.

**Comment by caprica** [comment]
It looks like this is the situation...

When dealing with media that creates sub-items, like a cue sheet, an m3u playlist, YouTube etc, if you play that media it will automatically play sub-items. This is actually quite useful.

If you do not want to play the sub-items, then instead of playing the media, use parse - this creates all the same sub-items and when you get the parsed status changed on the media you can play sub-item.

What we had previously was tied to the media player finished event - i.e. play the cue sheet and when it has finished it will create sub-items, you then play a sub-item. BUT, what happens now is that both VLC and your client application try and play a sub-item at the same time, leading to another crash.

So in summary, forget about auto-playing sub-items, use either play or parse to get the behaviour you want. Remember that for media you can either invoke play, or you can invoke prepare then parse.

**Comment by caprica** [comment]
As a result of #756, the media player SubItemService is somewhat sparse, it should be removed and its functionality moved elsewhere (maybe to the DefaultMediaPlayer class itself).

This will be tracked under a new ticket (pending).

**Comment by caprica** [comment]
Reminder to check if libvlc_media_duplicate could be useful.

**Comment by caprica** [comment]
What is media_list_add_file_content?

**Comment by caprica** [comment]
Should I deal in just a simple wrapper around the opaque libvlc_media_t for most things, and only create Media instances when they are actually needed?

**Comment by caprica** [comment]
It would have been nice to be able reuse a MediaList instance each time the subitems change. The idea was to use libvlc_media_list_set_media(list, media) as it was thought that when setting new media on an existing list it would automatically inherit the sub-items on that media. But this does NOT work.

Consequently it is not possible to reuse those MediaList instances.

**Comment by caprica** [comment]
This activity also needs to look at the various list component implementations, like AudioListPlayerComponent, EmbeddedMediaListPlayerComponent

**Comment by caprica** [comment]
This is something that may be revisited for vlcj-5.

**Comment by caprica** [comment]
These changes start at 2e1aae468601f5f627bd670ebfc0d4a91031be82 and run to 42db910dd2fb602446286d73a93f8d54ca70b043.

It is much better now, but as previously mentioned may be revisited for vlcj-5. It is a little but unfortunate that we have pairs of objects now - the component (e.g. Media) and the "ref" (e.g. MediaRef, representing a wrapper around the native resource). On the other hand, the intent of the API is clear, and who owns what should also be clear.

**Comment by caprica** [comment]
vlcj-5 changes will be considered under #765.

**Comment by caprica** [comment]
Some further work is required on MediaList, concerning convenience methods etc similar to #717.

**Comment by caprica** [comment]
Should be "finally" done with this now.
Changes 82df2131690eb3edf7bae8221ed1001013d5baed through e79848d04d17a63360e16b3ef97e426c3ee583f4.

**Comment by caprica** [comment]
This issue will not die.

Re-opening to deal with passing libvlc_media_t on MediaEventListener and MediaListEventListener and MediaPlayerEventListener, and MediaListPlayerEventListener.

**Comment by caprica** [comment]
Hopefully this is finally done at dcd8c48d4ff7d3edb2dab1570fcee6eb89e172f2.
